### PR TITLE
Add Restful API

### PR DESCRIPTION
**Pull request by AllentDan**
#213 
#214 
#189
#116

### CODE REVIEW

**Comment by lvhan028** [comment]
May update `profile_serving.py` to benchmark serving performance

File: docs/en/restful_api.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Restful API
+
+## Launch Service
```
Comment (lvhan028) [inline docs/en/restful_api.md:3]: All restful APIs are supposed to be presented in this guide

File: lmdeploy/serve/openai/protocol.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+
+class ModelPermission(BaseModel):
+    """Model permissions."""
+    id: str = Field(default_factory=lambda: f'modelperm-{shortuuid.random()}')
+    object: str = 'model_permission'
+    created: int = Field(default_factory=lambda: int(time.time()))
+    allow_create_engine: bool = False
+    allow_sampling: bool = True
+    allow_logprobs: bool = True
```
Comment (lvhan028) [inline lmdeploy/serve/openai/protocol.py:25]: Actually, turbomind doesn't support `logprobs`

File: docs/en/restful_api.md (hunk 1 above)
Comment (AllentDan) [reply docs/en/restful_api.md]: Added

File: lmdeploy/serve/openai/protocol.py (hunk 2 above)
Comment (AllentDan) [reply lmdeploy/serve/openai/protocol.py:25]: Yes, note this in the openai api docstring

File: benchmark/profile_restful_api.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+                                                 add_bos=False,
+                                                 add_eos=False)
+        return [len(token_ids) for token_ids in prompts_token_ids]
+
+
+def infer(server_addr: str, session_id: int, req_que: mp.Queue,
+          res_que: mp.Queue):
+    stats = []
+    while not req_que.empty():
+        prompt, input_seqlen, output_seqlen = req_que.get()
```
Comment (tpoisonooo) [inline benchmark/profile_restful_api.py:64]: rename `req_que` to `req_queue`,  `que` is not a common abbreviation.

File: benchmark/profile_restful_api.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        print(f'request info: session {session_id}, '
+              f'input_seqlen {input_seqlen}, output_seqlen {output_seqlen}')
```
Comment (tpoisonooo) [inline benchmark/profile_restful_api.py:66]: try `loguru.logger`,  it gives log time and help debugging in future.

File: docs/zh_cn/restful_api.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+```shell
+python lmdeploy/serve/openai/api_server.py ./workspace server_name server_port
+```
+
+然后用户可以打开 http://{server_name}:{server_port}/docs 详细查看所有的 API 及其使用方法。
+我们一共提供四个 restful api，其中三个仿照 OpenAI 的形式。不过，我们建议用户用我们提供的另一个 API: `generate`。
+它有更好的性能，提供更多的参数让用户自定义修改。
+
+### python
+
```
Comment (tpoisonooo) [inline docs/zh_cn/restful_api.md:16]: 加一点关于 swagger 的说明，例如哪个端口可以开 swagger，给一个链接如何把 swagger 接口转成 Rust/Go client.

**Review by tpoisonooo — APPROVED** [review]

File: benchmark/profile_restful_api.py (hunk 4 above)
Comment (AllentDan) [reply benchmark/profile_restful_api.py:66]: I'd like to. But this would introduce another third package in requirements. I don't know if it deserves it.

Comment (AllentDan) [reply benchmark/profile_restful_api.py:66]: @lvhan028

Comment (lvhan028) [reply benchmark/profile_restful_api.py:66]: Can we use `lmdeploy.utils.get_logger`?

File: docs/en/restful_api.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
+# Restful API
+
+### Launch Service
+
+```shell
+python lmdeploy/serve/openai/api_server.py ./workspace server_name server_port
```
Comment (lvhan028) [inline docs/en/restful_api.md:6]: Following this launching command, I got error like below:
```
```

Comment (AllentDan) [reply docs/en/restful_api.md]: Fixed

File: docs/en/restful_api.md [hunk 7]
```diff
@@ -108,5 +108,9 @@ curl http://{server_name}:{server_port}/v1/embeddings \
...
 ### FAQ
 
-When user got `"finish_reason":"length"` which means the session is too long to be continued.
-Please add `"renew_session": true` into the next request.
+1. When user got `"finish_reason":"length"` which means the session is too long to be continued.
+   Please add `"renew_session": true` into the next request.
+
+2. When OOM appeared at the server side, please reduce the number of `instance_num` when lanching the service.
+
+3. When the request with the same `instace_id` to `generate` got a empty return value and a negative `tokens`, please consider setting `sequence_start=false` for the second question and the same for the afterwards.
```
Comment (lvhan028) [inline docs/en/restful_api.md:116]: `instace_id` -> instance_id

File: docs/zh_cn/restful_api.md [hunk 8]
```diff
@@ -109,5 +109,9 @@ curl http://{server_name}:{server_port}/v1/embeddings \
...
 ### FAQ
 
-当返回结果结束原因为 `"finish_reason":"length"`，这表示回话长度超过最大值。
-请添加 `"renew_session": true` 到下一次请求中。
+1. 当返回结果结束原因为 `"finish_reason":"length"`，这表示回话长度超过最大值。
+   请添加 `"renew_session": true` 到下一次请求中。
+
+2. 当服务端显存 OOM 时，可以适当减小启动服务时的 `instance_num` 个数
+
+3. 当同一个 `instace_id` 的请求给 `generate` 函数后，出现返回空字符串和负值的 `tokens`，应该是第二次问话没有设置 `sequence_start=false`
```
Comment (lvhan028) [inline docs/zh_cn/restful_api.md:117]: instace_id -> instance_id

File: lmdeploy/serve/openai/api_server.py [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+def main(model_path: str,
+         server_name: str = 'localhost',
+         server_port: int = 23333,
+         instance_num: int = 32,
+         tp: int = 1):
+    """An example to perform model inference through the command line
+    interface.
+
+    Args:
+        model_path (str): the path of the deployed model
```
Comment (lvhan028) [inline lmdeploy/serve/openai/api_server.py:322]: mismatched argument sequence

File: benchmark/profile_restful_api.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+    token_throughput = np.sum(stats[:, 1], axis=0) / elapsed_time
+    req_throughput = n_req / elapsed_time
+
+    print(f'\n{"-" * 50}\nconcurrency: {concurrency}\n'
+          f'elapsed_time: {elapsed_time:.2f}s\n'
+          f'first_token latency(min, max, ave): '
+          f'{first_token_latency_min:.2f}s, {first_token_latency_max:.2f}s, '
+          f'{first_token_latency_ave:.2f}s\n'
+          f'token throughput: {token_throughput:.2f} token/s\n'
+          f'req throughput: {req_throughput} req/s\n'
```
Comment (lvhan028) [inline benchmark/profile_restful_api.py:204]: "req_throughput:.2f"

File: lmdeploy/turbomind/turbomind.py [hunk 11]
```diff
@@ -207,6 +208,13 @@ def _func(device_id, enque_output):
             t.start()
             self.threads[device_id] = t
 
+    async def async_stream_infer(self, *args, **kwargs):
+        """Async wrapper of self.stream_infer."""
+        for output in self.stream_infer(*args, **kwargs):
+            # Allow the pipeline add new requests into the queue.
+            await asyncio.sleep(0)
```
Comment (lvhan028) [inline lmdeploy/turbomind/turbomind.py:215]: Is `await` necessary?

Comment (AllentDan) [reply lmdeploy/turbomind/turbomind.py:215]: necessary here.

**Review by lvhan028 — APPROVED** [review]

**Comment by frankxyy** [comment]
有了这个基于fastapi的webserver，感觉是不是可以替换掉tritonserver了。。

**Comment by lvhan028** [comment]
是的。不过，TIS 我们暂时还不会删除

**Comment by renmada** [comment]
pipe = lmdeploy.pipeline(model_path,
                        backend_config=TurbomindEngineConfig(model_format='awq', tp=4))
```
RuntimeError: [TM][ERROR]  Assertion fail: [/lmdeploy/src/turbomind/triton_backend/transformer_triton_backend.cpp:32](http://192.168.51.13:7490/lmdeploy/src/turbomind/triton_backend/transformer_triton_backend.cpp#line=31)
```

**Pull request merged**